lms/services/canvas_api/authenticated.py

```diff
@@ -0,0 +1,140 @@
+"""Access to the authenticated parts of the Canvas API."""
+
+import marshmallow
+from marshmallow import fields
+
+from lms.services import CanvasAPIAccessTokenError
+from lms.validation import RequestsResponseSchema
+
+
+class TokenResponseSchema(RequestsResponseSchema):
```

Comment:
I think moving the file to the bottom is good.

The new S/O question about close collaborators and your suggestions for marking the files private help here too: 

* Everything in this directory is a close collaborator (the mutuality definition), so we don't need to get into the weeds about _how_ close they all are inside the bounded context, or worry about who is private to who within.
* By following your suggestion and marking the file as private, everything in it is private too. The boundary that matters (inside or out of the context) has been nicely established by that, and more clarification isn't required.
* Part of the definition is that all items within the boundary could happily live in one file if they were small enough, or many files as convenience dictates. So we are freed from having to read much into the location of the classes within the context. Wherever makes sense and fits is fine. In this case it makes sense to keep it close.
* Choosing to "zoom" in temporarily on one item makes it the unit under test, but does not change the relationship between the code. So we are freed from having to read into this zooming behavior.

In this case the extra tests weren't because I thought it was too big to test, just that the tests would be easier and shorter if I did it this way (I could have gone from the outside, but it's annoying). In this case I was just trying to catch a little coverage shortfall.

----
### An interesting distinction between our "close collaborators" and a domain model

I think this does bring up a reasonable point though, which helps clarify the collaborator rules.

It's pretty clear that the `AuthenticatedClient` concept creates it's own sub-domain which triggers some of these conclusions. You can, and I did choose to "zoom" into the `AuthenticatedClient` as a unit of test, and then into the `TokenResponseSchema` after that. In a pure domain driven model, each of these could be considered nested sub-domains, and our code layout reflects that. So:

 * This is perfectly reasonable definition of this domain `(*OtherStuff, (AuthenticatedClient, (TokenResponseSchema)))`
 * But we are defining our collaborators as `(*OtherStuff, AuthenticatedClient, TokenResponseSchema)`

So how do we get from the first to the second using our rules?

Part of our definition of a close collaborators is that aren't useful independently. The `AuthenticatedClient` doesn't work without the `TokenResponseSchema` and in the reverse it's useless. This gets us to:

 `(*OtherStuff, (AuthenticatedClient, TokenResponseSchema))`

The second step is a bit more tricky. There are plenty of other indicators that work, but I'd like to talk about:

> `Style`: Being a close collaborator is a mutual arrangement between a set of code. 

The crucial part here is that this is a `Style` rather than a `Guide`. We are to some extent stating this to be true when we use these words, rather than detecting this situation (don't be silly etc.).

So to go whole hog here we could have defined our domain model like this:

 * `(factory, (CanvasAPIClient, (AuthenticatedClient, BasicClient, TokenStore, (TokenResponseSchema))))`

But that isn't our definition of close collaborators, that's just our class layout / call stack. So it's an accurate view of the code, but it's not the close collaborator concept we're discussing.

* `(factory, CanvasAPIClient, AuthenticatedClient, BasicClient, TokenStore, TokenResponseSchema)`

I think you're file layout is on the money too, which is to say that both the factory and the client form the interface together.

----

### What a lot of waffle: what's your point?

I think drawing an intentional circle round the code and saying "inside, outside, doorway" and not getting into any more detail frees us from some of the discussions and distinctions here. This is perhaps the feature of doing this.

If you have a bounded context containing (by definition) mutually close collaborators you are always free to treat the code exactly as if it wasn't in a bounded context:

 * Mock in all the ways you would
 * Layout your code in all the ways you would
 * Separate and unit test as you would

We draw this boundary to give us additional freedoms:

 * To move / compose and decompose items within at will
 * To put things in convenient places
 * To not mock if we don't want to

And to do all of the above without being "bad" or creating a mess